Federated Wireless delivered the following key objects: CBSD, Grant, …

jni2000 opened the pull request:
Federated Wireless delivered the following key objects: CBSD, Grant and Domain Proxy. Please review. These objects are developed per earlier weekly review and discussion for setting up some common utility objects that can be used in other test cases. Please review.

kate-harrison commented on src/harness/sas_objects.py:
```diff
@@ -0,0 +1,297 @@
+    '''
+
+    def __init__(self,grantId,grantRequest):
+        self.GrantId = grantId
```
Our convention in this repository is to use `snake_case` for member variables.

jni2000 replied:
All the member variables used snake_case convention

kate-harrison commented on src/harness/sas_objects.py:
```diff
@@ -0,0 +1,297 @@
+    def __init__(self,grantId,grantRequest):
+        self.GrantId = grantId
+        self.GrantRequest = grantRequest
+        self.IsGrantAuthorisedInlastHeartbeat = False
```
Typo: "authorized" ("authorised" is the British spelling)

jni2000 replied:
authorized  used instead of authorised

kate-harrison commented on src/harness/sas_objects.py:
```diff
@@ -0,0 +1,297 @@
+        return hearbeatRequest
+
+    def UpdateGrantAuthorisedInlastHeartbeat(self,IsAuthorized):
+        IsGrantAuthorisedInlastHeartbeat = IsAuthorized
```
This doesn't actually set the member variable.

jni2000 replied:
used the self.grant_authorized_in_last_heartbeat instead of grant_authorized_in_last_heartbeat

kate-harrison commented on src/harness/sas_objects.py:
```diff
@@ -0,0 +1,297 @@
+    def GetRequestOperationParam(self):
+        return self.GrantRequest['operationParam']
+
+    def ConstructHeartBeatRequest(self):
```
Typo: "Heartbeat" (it is only one word)

jni2000 replied:
changed to heartbeat

kate-harrison commented on src/harness/sas_objects.py:
```diff
@@ -0,0 +1,297 @@
+    '''
+    def __init_(self,cbsdId,successfulRegRequest,grantId,granRequests):
+        self.CbsdId = cbsdId
+        self.GrantObjects = {"":Grant}
```
This appears to be incomplete.

jni2000 replied:
used the empty dictionary , item will be inserted later

kate-harrison commented on src/harness/sas_objects.py:
```diff
@@ -0,0 +1,297 @@
+		Construct heartbeat request for all active grants
+		Return operating params for all grants wh{"grantId":"grantRequest"}ich are authorized to transmit
+    '''
+    def __init_(self,cbsdId,successfulRegRequest,grantId,granRequests):
```
Typo: "grantRequests"

jni2000 replied:
fixed the typo

kate-harrison commented on src/harness/sas_objects.py:
```diff
@@ -0,0 +1,297 @@
+    def GetRegistrationRequest(self,grantId):
+        return self.SuccessfulRegRequest
+
+    def IsCbsdIdhasAtleastOneGrant(self):
```
This would more naturally be called "CbsdHasAtLeastOneGrant" -- the current name is not proper English.

jni2000 replied:
used the suggested name for the function

kate-harrison commented on src/harness/sas_objects.py:
```diff
@@ -0,0 +1,297 @@
+
+    def IsCbsdIdhasAtleastOneGrant(self):
+        for grantId,grantRequest in self.GrantObjects.items():
+            if grantRequest['operationState'] == "GRANTED":
```
This is not the right check.

jni2000 replied:
checking the authorized flag

kate-harrison commented on src/harness/sas_objects.py:
```diff
@@ -0,0 +1,297 @@
+                return True
+        return False
+
+    def ConstructSingleHeartBeatRequest(self,cbsdId,grantId):
```
This seems like a duplicate of the function in the Grant class.

jni2000 replied:
removed the duplicate function

kate-harrison commented on src/harness/sas_objects.py:
```diff
@@ -0,0 +1,297 @@
+         return hearbeatRequest
+
+    def ConstructHeartBeatRequestForAllActiveGrant(self):
+        heartBeatRequests = ""
```
This needs to be a list, not a string.

jni2000 replied:
used the list instead of string

kate-harrison commented on src/harness/sas_objects.py:
```diff
@@ -0,0 +1,297 @@
+        for grantId,granObject in self.GrantObjects.items():
+            if granObject.IsGrantAuthorisedInlastHeartbeat:
+             OperationParam = granObject['operationParam']
+             OperationParams.insert(OperationParam)
```
This function takes two arguments.

jni2000 replied:
used append instead of insert

jni2000 commented:
We have finished revision on your comments, please take a second review. Also, we have done some minimal testing as much as we can.

kate-harrison commented on src/harness/sas_objects.py:
```diff
@@ -0,0 +1,353 @@
+#    Copyright 2016 SAS Project Authors. All Rights Reserved.
```
2018 now

jni2000 replied:
Updated   to 2018

kate-harrison commented on src/harness/sas_objects.py:
```diff
@@ -0,0 +1,353 @@
+    Retrieve was authorized in last heartbeat
+    """
+
+    def __init__(self, g_id, g_request):
```
Please do not use abbreviations in the input variables -- there's no benefit to saving space and it's less readable this way.

jni2000 replied:
Abbreviations   are changed to meaningful names

kate-harrison commented on src/harness/sas_objects.py:
```diff
@@ -0,0 +1,353 @@
+class Grant:
+
+    """
+    A Grant object has the following methods
```
This is likely to fall out of date if/when we add new methods. Either remove this docstring or replace with a high-level description of how the class should be used and what it's intending to accomplish.

jni2000 replied:
Added   high level description for all the classes.

kate-harrison commented on src/harness/sas_objects.py:
```diff
@@ -0,0 +1,353 @@
+        Construct heartbeat request for all active grants
+        Return operating params for all grants which are authorized to transmit
+    """
+    def __init__(self, c_id, reg_request, g_id, grant_requests):
```
Please spell out words rather than using non-standard abbreviations.

jni2000 replied:
Abbreviations   are changed to meaningful names

kate-harrison commented on src/harness/sas_objects.py:
```diff
@@ -0,0 +1,353 @@
+        Construct heartbeat request for all active grants
+        Return operating params for all grants which are authorized to transmit
+    """
+    def __init__(self, c_id, reg_request, g_id, grant_requests):
```
`grant_ids` should be plural.

jni2000 replied:
Updated   to plural

kate-harrison commented on src/harness/sas_objects.py:
```diff
@@ -0,0 +1,353 @@
+        """
+        return self.successful_registration_request
+
+    def cbsd_id_has_atleast_one_grant(self):
```
"at least" is two words. A CBSD has a Grant, not a CBSD ID. 

Based on the body of the function, this is retrieving only authorized Grants. If that is the intention, a more appropriate name would be `cbsd_has_at_least_one_authorized_grant`.

However, I think we need a function that checks to see whether or not the CBSD has any *active* (i.e. non-terminated/non-relinquished) Grants. This will allow us to quickly pass over any CBSDs which no longer have Grants in some of our other functions.

jni2000 replied:
Fixed as   per suggestion

kate-harrison commented on src/harness/sas_objects.py:
```diff
@@ -0,0 +1,353 @@
+        """
+        return self.cbsd_id
+
+    def get_grant_request(self, grant_id):
```
This is getting the Grant object, not a grant request.

jni2000 replied:
Fixed as   per suggestion

kate-harrison commented on src/harness/sas_objects.py:
```diff
@@ -0,0 +1,353 @@
+    def get_grant_request(self, grant_id):
+        """
+
+        :rtype: grant_objects
```
Plural is incorrect here.

jni2000 replied:
Fixed as   per suggestion

kate-harrison commented on src/harness/sas_objects.py:
```diff
@@ -0,0 +1,353 @@
+    def get_registration_request(self):
+        """
+
+        :rtype: successful_registration_request
```
`registration_request` will probably make more sense to callers.

jni2000 replied:
Fixed as   per suggestion

kate-harrison commented on src/harness/sas_objects.py:
```diff
@@ -0,0 +1,353 @@
+                return True
+        return False
+
+    def construct_heartbeat_request_for_all_active_grant(self):
```
There should be an "s" at the end.

jni2000 replied:
Fixed as   per suggestion

kate-harrison commented on src/harness/sas_objects.py:
```diff
@@ -0,0 +1,353 @@
+    def construct_heartbeat_request_for_all_active_grant(self):
+        """
+
+        :rtype: heartbeat_requests
```
"list of heartbeat requests" would be more descriptive and accurate.

jni2000 replied:
Fixed as   per suggestion

kate-harrison commented on src/harness/sas_objects.py:
```diff
@@ -0,0 +1,353 @@
+        :rtype: heartbeat_requests
+        """
+        heartbeat_requests = []
+        for grant_id, grant_object in self.grant_objects.items():
```
You can use `_` instead of `grant_id` since the Grant ID is not used. Even better is to iterate over `self.grant_objects.values()` directly.

jni2000 replied:
Used the   values directly instead of items

kate-harrison commented on src/harness/sas_objects.py:
```diff
@@ -0,0 +1,353 @@
+            heartbeat_requests.append(heartbeat_request)
+        return heartbeat_requests
+
+    def get_all_operation_params_of_grant_request(self):
```
Please fix the title of this method to reflect the fact that it's only retrieving operationParams for Grants that are authorized.

jni2000 replied:
Added a   title to explain the function

kate-harrison commented on src/harness/sas_objects.py:
```diff
@@ -0,0 +1,353 @@
+        :rtype: operation_params
+        """
+        operation_params = []
+        for grant_id, grant_object in self.grant_objects.items():
```
Again, you can use `self.grant_objects.values()` instead.

jni2000 replied:
Fixed as   per suggestion

kate-harrison commented on src/harness/sas_objects.py:
```diff
@@ -0,0 +1,353 @@
+        """
+        operation_params = []
+        for grant_id, grant_object in self.grant_objects.items():
+            if True == grant_object.get_grant_authorized_in_last_heartbeat():
```
`True ==` is redundant here.

jni2000 replied:
Fixed as   per suggestion

kate-harrison commented on src/harness/sas_objects.py:
```diff
@@ -0,0 +1,353 @@
+        # checking for the number of registration requests matches number of grant requests
+        if len(registration_requests) != len(grant_requests):
+            return
+        self.dp_certificate = certificate
```
Why add the `dp_` prefix?

jni2000 replied:
It is   not really needed. Removed it now.

kate-harrison commented on src/harness/sas_objects.py:
```diff
@@ -0,0 +1,353 @@
+        self.dp_key = key
+        self.cbsd_objects = {}
+        cbsd_id = []
+        self.sas_admin = sas.GetTestingSas()
```
This may be better as an input parameter.

Moreover, the output of this function is a tuple (not a singular value) so line 191 will fail.

jni2000 replied:
No we   are taking the calling test case object as an input so that _sas and   _sas_admin can be used and also the methods like assertRegistered can be   invoked.


Changed   the input parameter to the calling test case object.

kate-harrison commented on src/harness/sas_objects.py:
```diff
@@ -0,0 +1,353 @@
+        self.cbsd_objects = {}
+        cbsd_id = []
+        self.sas_admin = sas.GetTestingSas()
+        for registerRequest in registration_requests[:]:
```
Use snake_case `registration_request` instead of CamelCase `registerRequest` for consistency.

jni2000 replied:
Changed   to register_request

kate-harrison commented on src/harness/sas_objects.py:
```diff
@@ -0,0 +1,353 @@
+        self.cbsd_objects = {}
+        cbsd_id = []
+        self.sas_admin = sas.GetTestingSas()
+        for registerRequest in registration_requests[:]:
```
Use of `[:]` is completely unnecessary.

jni2000 replied:
Removed

kate-harrison commented on src/harness/sas_objects.py:
```diff
@@ -0,0 +1,353 @@
+        self.dp_certificate = certificate
+        self.dp_key = key
+        self.cbsd_objects = {}
+        cbsd_id = []
```
Use `cbsd_ids` plural since this is a list.

jni2000 replied:
Changed   as suggested

kate-harrison commented on src/harness/sas_objects.py:
```diff
@@ -0,0 +1,353 @@
+            response = self.sas_admin.Registration(registerRequest, self.dp_certificate, self.dp_key)\
+                                                                  ['registrationResponse'][0]
+            cbsd_id.append(response['cbsdId'])
+            self.sas_admin.assertEqual(response['response']['responseCode'], 0)
```
This makes no sense whatsoever -- the function `assertEqual` is part of the testcase framework and unittest class, and does not belong to the SAS Admin interface.

I'm going to stop reviewing at this point until the code is internally tested.

jni2000 replied:
This is   corrected. Thanks for pointing it out. We have changed the input parameter to   the invoking sas_testcase object and used that object for assert methods.

jni2000 commented:
Kate and all reviewers: 

I want to apologize here that I did not do better job on controlling the code quality and let such obvious errors occurred again. I will take immediate action to enforce our internal code quality control and review process to avoid such situations. As for the technical aspects and convention style related, please allow us to learn the existing codes and gradually adopt them, we appreciate you comments on these as well.  Thanks and we are taking action on making corrections and do thorough internal review..

jni2000 commented:
Thanks to the review comments and suggestions. We have reworked the codes, done internal review and have tested the codes as much as we can. Please continue review.

jni2000 commented:
We have submitted changes responding the review comments 6 days before, please move on reviewing our revision. Thanks.

jni2000 commented:
Please can we continue and finish up the review of this? It has been there for 24 days and this blocks the MCP.1 development.

kate-harrison commented:
At your request, I have been concentrating on review and development of the protocol test cases. Now that those are (nearly) done, I'm happy to prioritize according to your suggestions.

Please note that other companies are welcome to review this PR; I do not need to be the main reviewer.

And in general, please refrain from adding the "reviewer:google" tag even if someone from Google makes comments on a PR -- we will add it ourselves if we wish to be a blocking reviewer.

kate-harrison commented on src/harness/sas_objects.py:
```diff
@@ -0,0 +1,387 @@
+    def initialize(self, registration_requests, grant_requests):
+        """
+        Args
+            registration_requests:  A list of dictionary with a single key-value pair
```
Unwrap.

jni2000 replied:
As per   suggestion used the non wrapped version of grant requests

kate-harrison commented on src/harness/sas_objects.py:
```diff
@@ -0,0 +1,387 @@
+        self.cbsd_objects = {}
+        self.testcase = testcase
+
+    def initialize(self, registration_requests, grant_requests):
```
Why is this separate from `__init__()`?

jni2000 replied:
This   was done to use the same DP object for a different set of CBSDs and grants.   This is to be used in the MCP test case where the DP object is used to  in different iterations for different set   of CBSDs. Hence the function to add CBSDs and grants to the DP object is   separated from the constructor.

kate-harrison commented on src/harness/sas_objects.py:
```diff
@@ -0,0 +1,387 @@
+            grant_response = self.testcase._sas.Grant(grant_request, self.ssl_cert, self.ssl_key)[
+                'grantResponse'][0]
+            if grant_response['response']['responseCode'] == 0:
+                cbsdobject = Cbsd(cbsd_ids[index], registration_requests[index],
```
Please fix the variable name.

jni2000 replied:
fixed   the variable name

kate-harrison commented on src/harness/sas_objects.py:
```diff
@@ -0,0 +1,387 @@
+                cbsdobject = Cbsd(cbsd_ids[index], registration_requests[index],
+                                  [grant_response['grantId']], [grant_request])
+                self.cbsd_objects.update(
+                    {grant_response['cbsdId']: cbsdobject})
```
It will probably be more readable if you use either `cbsd_id[index]` or `cbsd_object.cbsdId()`.

jni2000 replied:
Logic   has been changed

kate-harrison commented on src/harness/sas_objects.py:
```diff
@@ -0,0 +1,387 @@
+            if grant_response['response']['responseCode'] == 0:
+                cbsdobject = Cbsd(cbsd_ids[index], registration_requests[index],
+                                  [grant_response['grantId']], [grant_request])
+                self.cbsd_objects.update(
```
Why are you using `update`?

jni2000 replied:
Logic   has been changed

kate-harrison commented on src/harness/sas_objects.py:
```diff
@@ -0,0 +1,387 @@
+    def getCbsdsWithAtLeastOneAuthorizedGrant(self):
+        """
+        returns:
+            cbsd_objects: list of cbsd objects
```
No need to give the var name.

jni2000 replied:
removed   var name

kate-harrison commented on src/harness/sas_objects.py:
```diff
@@ -0,0 +1,387 @@
+        """
+        cbsd_objects = []
+        for cbsd_object in self.cbsd_objects.values():
+            for grant_object in cbsd_object.grant_objects.values():
```
You already have a function (currently called `doesCbsdHasActiveGrant`) to do this. Use it.

jni2000 replied:
Used as   per suggestion

kate-harrison commented on src/harness/sas_objects.py:
```diff
@@ -0,0 +1,387 @@
+            assert(not "Invalid grant response code")
+
+
+    def performHeartbeatAndUpdateGrants(self):
```
The name (and especially the comment) needs to give the reader a better sense of what this function does. Right now it's nearly impossible to tell, even after reading the docstring.

jni2000 replied:
This function issues heartbeat requests for all active grants and do appropriate grant adjustments based on the heartbeat response instructions. Would you please suggest a better/more descriptive name for this ? Thanks.

kate-harrison commented on src/harness/sas_objects.py:
```diff
@@ -0,0 +1,387 @@
+        """
+            This function construct heartbeat request for all active grants and updates the grants
+        """
+        for cbsd_object_item in self.cbsd_objects.values():
```
You need to HB for all Grants at the same time, not once per CBSD/Grant pair. The current implementation is completely antithetical to the concept of a domain proxy.

jni2000 replied:
Fixed.

kate-harrison commented on src/harness/sas_objects.py:
```diff
@@ -0,0 +1,387 @@
+                    heartbeat_request, self.ssl_cert, self.ssl_key)['heartbeatResponse'][0]
+                grant_object = cbsd_object_item.grant_objects[heartbeat_response['grantId']]
+                if heartbeat_response['response']['responseCode'] == 0:
+                    transmit_expire_time = datetime.strptime(heartbeat_response['transmitExpireTime'],
```
We don't need to do these checks: it's not a protocol test case. So in fact you can delegate all of the logic to `mapResponseCodeToGrantState`.

jni2000 replied:
Removed   as per suggestion

kate-harrison commented on src/harness/sas_objects.py:
```diff
@@ -0,0 +1,387 @@
+        """
+            This function maps the heartbeat response code to the grant state
+        """
+        if (heartbeat_response_code == 502):
```
The behavior should be as follows:

0: authorized_in_last_heartbeat -> True
500: is_terminated -> True, authorized_in_last_heartbeat -> False
501: authorized_in_last_heartbeat -> False
502: relinquish Grant (send a request to the SAS), is_terminated -> True

jni2000 replied:
Updated   as per suggestion. The success case (0)    and the 502 case to relinquish the grant is handled in the calling   function other values updated as per suggestion.

kate-harrison commented on src/harness/sas_objects.py:
```diff
@@ -0,0 +1,387 @@
+                        }
+                        self.testcase._sas.Relinquishment(
+                            relinquish_request, self.ssl_cert, self.ssl_key)
+                        heartbeat_responses_with_operation_parameters.update(
```
Why `update`?

jni2000 replied:
Logic   has been changed

kate-harrison commented on src/harness/sas_objects.py:
```diff
@@ -0,0 +1,387 @@
+                    #If any heartbeats contain suggested operating params,
+                    #relinquish the corresponding grant
+                    if 'operationParam' in heartbeat_response:
+                        relinquish_request = {
```
Maybe the Grant object should help to construct the relinquishment request.

jni2000 replied:
As both   cbsdId and grantId are required, we used heartbeat response for constructing   the relinquishment in this case. Hope it is ok.

kate-harrison commented on src/harness/sas_objects.py:
```diff
@@ -0,0 +1,387 @@
+                                'grantId': grant_object.getGrantRequest()['grantId']
+                            }]
+                        }
+                        self.testcase._sas.Relinquishment(
```
You probably want to do this in batch as well.

jni2000 replied:
Updated   accordingly.

kate-harrison commented on src/harness/sas_objects.py:
```diff
@@ -0,0 +1,387 @@
+            #For all grants in the previous step, request a new grant with the
+            #suggested operation parameters
+            for cbsd_id, heartbeatResponse in heartbeat_responses_with_operation_parameters.items():
+                grant_0 = json.load(
```
No need to load, since you're overwriting all of the values anyway.

jni2000 replied:
changed

kate-harrison commented on src/harness/sas_objects.py:
```diff
@@ -0,0 +1,387 @@
+                    'grantResponse'][0]
+                self.cbsd_objects.grant_objects.update(
+                    {response['grantId']: request })
+                if response['response']['responseCode'] == "0":
```
The response code is NOT a string.

jni2000 replied:
changed

kate-harrison commented on src/harness/sas_objects.py:
```diff
@@ -0,0 +1,387 @@
+                request = {'grantRequest': [grant_0]}
+                response = self.testcase._sas.Grant(request , self.ssl_cert, self.ssl_key)[
+                    'grantResponse'][0]
+                self.cbsd_objects.grant_objects.update(
```
At this point you do not know that the grant request was successful, and you cannot assume this.

jni2000 replied:
Logic   has been changed

kate-harrison commented on src/harness/sas_objects.py:
```diff
@@ -0,0 +1,387 @@
+                    {response['grantId']: request })
+                if response['response']['responseCode'] == "0":
+                    #Heartbeat all active grants
+                    self.heartbeatRequestForAllGrants()
```
Why is it this inside the `for` loop? Please check the indentation level of this and the previous code.

jni2000 replied:
Logic   has been changed

kate-harrison commented on src/harness/sas_objects.py:
```diff
@@ -0,0 +1,387 @@
+        """
+            This function performs heartbeat request for all grants
+        """
+        for cbsd_object_item in self.cbsd_objects.values():
```
Send as a batch, not individually. You are ignoring the whole point of the domain proxy.

jni2000 replied:
Updated   accordingly.

kate-harrison commented on src/harness/sas_objects.py:
```diff
@@ -0,0 +1,387 @@
+                    heartbeat_request, self.ssl_cert, self.ssl_key)['heartbeatResponse'][0]
+                grant_object = cbsd_object_item.grant_objects[heartbeat_response['grantId']]
+                if heartbeat_response['response']['responseCode'] == 0:
+                    transmit_expire_time = datetime.strptime(heartbeat_response['transmitExpireTime'],
```
No need.

jni2000 replied:
removed

kate-harrison commented on src/harness/sas_objects.py:
```diff
@@ -0,0 +1,387 @@
+
+    def mapResponseCodeToGrantState(self, heartbeat_response_code, grant_object):
+        """
+            This function maps the heartbeat response code to the grant state
```
This function does more than this: it also updates the grant object. You need to declare side effects in the docstring of the function.

jni2000 replied:
Updated   as per suggestion

TaranFurnell commented on src/harness/sas.py:
```diff
@@ -203,6 +208,42 @@ def _GetDefaultSasSSLCertPath(self):
+                 ssl_cert if ssl_cert else self._GetDefaultSasSSLCertPath(),
+                 ssl_key if ssl_key else self._GetDefaultSasSSLKeyPath())
+
+  def _DownloadFile(self, url, ssl_cert, ssl_key):
```
What is this function supposed to achieve that _RequestGet does not already do? It appears to be the same function but with some values hard-coded. This includes the ciphers which is something the security tests need to modify (hence the use of TlsConfig)

jni2000 replied:
This   portion of code has been removed from this PR as Google will take over the   implementation of the FAD Object.

jni2000 commented on src/harness/sas_objects.py:
```diff
@@ -0,0 +1,387 @@
+        """
+        Args:
+           grant_id: grant id of the request
+           grant_request: A dictionary with a single key-value pair where the key is
```
As per   suggestion used the non wrapped version of grant requests

jni2000 commented on src/harness/sas_objects.py:
```diff
@@ -0,0 +1,387 @@
+        return self.grant_request
+
+    def getRequestOperationParam(self):
+        return self.grant_request['grantRequest'][0]['operationParam']
```
Handled   accordingly

jni2000 commented on src/harness/sas_objects.py:
```diff
@@ -0,0 +1,387 @@
+        heartbeat_request = {
+            'heartbeatRequest': [{
+                'grantId': self.grant_id,
+                'cbsdId': self.grant_request['grantRequest'][0]['cbsdId'],
```
Handled   accordingly

jni2000 commented on src/harness/sas_objects.py:
```diff
@@ -0,0 +1,387 @@
+            'heartbeatRequest': [{
+                'grantId': self.grant_id,
+                'cbsdId': self.grant_request['grantRequest'][0]['cbsdId'],
+                'operationState': "GRANTED"
```
Used   single quotation mark

jni2000 commented on src/harness/sas_objects.py:
```diff
@@ -0,0 +1,387 @@
+    def setState(self, state):
+        self.state = state
+
+    def isGrantAuthorizedInLastHeartbeat(self):
```
As per   suggestion updated grant state logic handling

jni2000 commented on src/harness/sas_objects.py:
```diff
@@ -0,0 +1,387 @@
+                    cbsd_objects.append(cbsd_object)
+        return cbsd_objects
+
+class FullActivityDump():
```
Removed   as per suggestion

jni2000 commented on src/harness/sas_objects.py:
```diff
@@ -0,0 +1,387 @@
+
+class Cbsd():
+    """
+        Holds the Cbsd related parameters
```
Updated   accordingly.

jni2000 commented on src/harness/sas_objects.py:
```diff
@@ -0,0 +1,387 @@
+        """
+        Args:
+            cbsd_id:  cbsd id of the device extracted from the registration response
+            registration_request: A dictionary with a single key-value pair where the key is
```
As per   suggestion used the non wrapped version of grant requests

jni2000 commented on src/harness/sas_objects.py:
```diff
@@ -0,0 +1,387 @@
+            registration_request: A dictionary with a single key-value pair where the key is
+            "registrationRequest" and the value is a registration request content
+            grant_ids: list of grant ids
+            grant_requests: A list of dictionary with a single key-value pair where the key is
```
As per   suggestion used the non wrapped version of grant requests

jni2000 commented on src/harness/sas_objects.py:
```diff
@@ -0,0 +1,387 @@
+        self.cbsd_id = cbsd_id
+        self.grant_objects = {}
+        index = 0
+        for grant_request in grant_requests:
```
As per   suggestion used zip

jni2000 commented on src/harness/sas_objects.py:
```diff
@@ -0,0 +1,387 @@
+    def getRegistrationRequest(self):
+        return self.registration_request
+
+    def doesCbsdHasActiveGrant(self):
```
Updated   as per suggestion

jni2000 commented on src/harness/sas_objects.py:
```diff
@@ -0,0 +1,387 @@
+
+    def doesCbsdHasActiveGrant(self):
+        for grant_object in self.grant_objects.values():
+            return grant_object.isGrantActive()
```
fixed

jni2000 commented on src/harness/sas_objects.py:
```diff
@@ -0,0 +1,387 @@
+        else:
+            return False
+
+    def isGrantActive(self):
```
Updated   as per suggestion

jni2000 commented:
Kate and Taran

We have addressed all your comments. Please continue. Tow points to note here:

1. For comment on the naming of "performHeartbeatAndUpdateGrants(self)", please can you suggest a better or more descriptive name ?
2. For comment "Grant object should help to construct the relinquishment request" on line 256 of sas_objects.py, since both cbsdId and grantId are required and available in the heartbeat response, we used heartbeat response for constructing the relinquishment in this case. Let us know if this is significantly inappropriate and we will make the change.

jni2000 commented:
** Review comments from Takashi and responses from FW **

Maybe I need some more background. 

[1] I’m confused with the name “sas_object”. 
It took me time to understand what it is. 
It is used in testcase.py of MCP, right?  It looks an object of CBSD related information, not SAS. 
Name is misleading. (I first thought it is used in the fake_sas…) 

[jni2000] Yes. It is used mainly in MCP and related test cases. It implements multiple objects (Grant, Cbsd and DomainProxy). FullActivityDump object which was originally part of this has been removed now. The name is somewhat misleading. We will change it to “test_harness_objects.py”.

[2] hasActiveGrant (self):
    Term “active” is confusing. 
    We need a definition in the code otherwise please consider to change the term to prevent confusion. 
-It means “Non-Terminated” (=Approved), right?
     -It doesn’t mean “Authorized”, right?
[jni2000] yes. It means non-terminated. It may not be ‘Authorized’. The function name was chosen as per the F2F document and recommendation from other reviewers. We will add comments in the code to clarify.

[3] class Grant():
 > """Holds the Grant request parameters"""
Grant Response should be also included, right? Or “Grant request + only grantId”? 
 I don’t understand which is the aim of the code. 
[jni2000] This is implemented as per the wg4 F2F document. I think the Grant and Cbsd objects mainly assist the DomainProxy object for bulk messaging. Only successful registration and grant procedures will have the respective Cbsd and Grant objects created.

[4] class Cbsd():
  >"""Holds the CBSD related parameters"""
  >    registration_request:A registrationRequest object as defined in the SAS-CBSD TS.

　　What is “registration_request”?
　Is it Request message itself or merged version with RegCondx (Preload)?
　　i.e. ‘registration field’ in the “CBSD Record” of the dump file?
[jni2000] “registration_request” is the registrationRequest dictionary as defined in the SAS-CBSD TS. It is NOT the ‘registration” field in the CbsdData format (which is the “CBSD Record” of the dump file) defined in SAS-SAS TS. It does not include the Reg Conditionals. The conditionals will have to be preloaded separately. It contains only the parameters that needs to be sent in the registration request by a Domain Proxy.

[5] class DomainProxy():
(a)What is “registration_requests”?
 I thought an array of “class Cbsd():” may be easier to use?
[jni2000] Actually it is the other way. The Cbsd objects are built by the Domain Proxy object using the registration_requests and not that the Domain proxy object is created from a list of Cbsd() objects.

　   (b)How can grantId be obtained? 
      >grant_responses = self.testcase._sas.Grant(grant_requests_wrap, self.ssl_cert, self.ssl_key)['grantResponse']
          I cannot find [‘grantResponse’] in the Grant object. 
[jni2000] Actually this invokes the method named ‘grant’ in the class “SasImpl” which sends the grant requests to the SAS UUT and returns the response message from the SAS UUT. The grantIds in the response are used.

jni2000 commented:
> [4] class Cbsd():
I don’t understand what you mean in your reply,
Do you mean CBSD() stores only required parameters except REG-Conditionals? 
Registration request contains REG-Conditional parameters in some cases, but. 
do you mean that MCP.1 test will be performed to limit such that 
REG-Conditional parameters will be ALWAYS preloaded? 

[jni2000]  The Cbsd() object store only the parameters that are sent in the registration request. If some CBSD has REG-Conditionals or CPI information parameters to be sent in the registration request than those will also be stored on the Cbsd(). Basically the Cbsd() stores the contents of the RegistrationRequest Object as defined in section 10.1.1 (Table4) of the SAS-CBSD spec. If any CBSD registration requires the REG-conditionals to be pre-loaded then that step has to be done prior to using a DomainProxy() Object to initiate registration procedure.
In MCP.1 test case each iteration has a set of CBSDs to be registered using a DomainProxy. The registration parameters for these CBSDs and the conditionals to be pre-loaded are provided in the configuration file as separate parameters. The conditionals are injected into the SAS UUT by invoking the admin API and the registration request message contents are used to create the DomainProxy Object. Then using the DomainProxy Object we initiate the registration, grant and heartbeat for these CBSDs. So injecting the conditionals and performing registration are done in two different steps. Hope this answers the question.


> [5]class DomainProxy(): 
I realize that it seems that self.testcase._sas.Grant() and Grant() are completely different, right? 
(both are in the sas_object.py)
Based on that, I’m looking into the code again, and I have further comments/questions. 
c)comment is confusing. 
sas_object.py line111>class DomainProxy():
sas_object.py line112>  """Holds the domain proxy related parameters"""
Actually DomainProxy() class is not a storage, but this class contains procedure tools. 
So, this comment is confusing. 

[jni2000] Yes, the DomainProxy() class contains procedure tools as well. We will update the comment as 

“”” Performs Domain Proxy related operations.

This class stores Domain Proxy related information like the SSL certificate, key to be used and the set of
CBSD objects belonging to this Domain Proxy. Bulk operations like registration, grant and heartbeat
procedures are performed on all of the CBSDs belonging to this Domain Proxy.
“””

>    d)There is a method called initialize() in this class. 
sas_object.py line133>  def initialize(self, registration_requests, grant_requests):
  How is this used?   
grant_requests cannot be given before registration. 
e.g. cbsdId in the grant_requests should be edited after the registration. 
  What is the difference between similar method in the sas_testcase.py?
sas_testcase.py>  def assertRegisteredAndGranted(self, registration_request, grant_request,
sas_testcase.py>                                 conditional_registration_data=None):

[jni2000] The initialize method does the following
●	Send N registration requests, expects success for all
●	Send N grant requests, allows failure
●	Make one CBSD object per successful grant request
The grant requests contain frequency and related information for a CBSD. The cbsdId field is added into the grantRequest after the registration is successful and the cbsdId returned in the corresponding registration Request is used. The initialize() method is used to register a set of CBSDs and a request a grant for each of them using the DomainProxy credentials.

The function assertRegisteredAndGranted in sas_testcase.py is similar but it does not allow for failures of grants. If a test case uses assertRegisteredAndGranted function then it is the responsibility of the test case to store the return values (cbsdId and grantId) and use them appropriately in subsequent procedures like heartbeat. DomainProxy Object is used to simplify these process, mainly for MCP.1 test case.

> Best Regards,
Takashi 

[jni2000]: I will send you a copy of WG4 test code development design working document which shared by Kate a while ago for your reference. Also available here: https://docs.google.com/document/d/1K6G-SUE0LAk3EUfNIf5Vb7TDREWshHPm6ZVd50ufd_s/edit#heading=h.t2e1nkvciff0

jni2000 commented:
Dear all reviewers:

Can we please wrap up the review on this PR by today if no more concerns ? We need to move on to MCP.1 functional portion implementation and this will unblock a lot of other function test cases implementation. Thanks.

TaranFurnell commented on src/harness/test_harness_objects.py:
```diff
@@ -0,0 +1,299 @@
+    for heartbeat_response, heartbeat_request in zip(heartbeat_responses, heartbeat_requests):
+        if heartbeat_response['response']['responseCode'] != 0:
+          # If any heartbeats contain suggested operation params,construct relinquish requests
+          if 'operationParam' in heartbeat_response:
```
if response code is TERMINATED_GRANT and operationParam is not present then raise or log an error. As the response is invalid.

jni2000 replied:
Added   error log as per suggestion

TaranFurnell commented on src/harness/test_harness_objects.py:
```diff
@@ -0,0 +1,299 @@
+        if heartbeat_response['response']['responseCode'] != 0:
+          # If any heartbeats contain suggested operation params,construct relinquish requests
+          if 'operationParam' in heartbeat_response:
+            relinquish_request = {
```
TERMINATED_GRANT responses do not make a relinquishment request, they proceed straight to making a grant request.

jni2000 replied:
In case   of terminated grant response code only grant request is sent

TaranFurnell commented on src/harness/test_harness_objects.py:
```diff
@@ -0,0 +1,299 @@
+      relinquish_requests_wrap = {
+        'relinquishmentRequest': relinquish_requests
+      }
+      self.testcase._sas.Relinquishment(relinquish_requests_wrap,
```
If any relinquishments fail it is most likely an issue with the test code (assuming the SAS UUT passes the basic protocol tests), regardless the tests and SAS UUT are in an inconsistent state so best to log what happened, and raise an error to stop the test execution (the test results would not be valid anymore anyway).

jni2000 replied:
Error   logged in case of  Relinquishment   fails

TaranFurnell commented on src/harness/test_harness_objects.py:
```diff
@@ -0,0 +1,299 @@
+  def mapResponseCodeToGrantState(self, heartbeat_response_code, grant_object):
+    """Maps the heartbeat response and update the grant state"""
+    if heartbeat_response_code == 502:
+      grant_object.is_terminated = True
```
If an UNSYNC_OP_PARAM response code occurs it's probably worth at least logging an error. Removing this check and implement a check at the bottom for any non-matching responses (as comment on line 221 suggests) should be sufficient.

jni2000 replied:
Error   logged in case of  Response code   UNSYNC_OP_PARAM

jni2000 commented:
Taran

Thanks for the review comments. We are working on responses to them right now. At the meantime can you please check if these are the only ones we need to address as of now ? We would like to address them all in once if possible so that we can speed up the process, as this one is what MCP.1 and other test cases depends on. Please help on this. Thanks

James

jni2000 commented on src/harness/sas.py:
```diff
@@ -164,6 +164,11 @@ def _SasRequest(self, method_name, request, ssl_cert=None, ssl_key=None):
                            ssl_cert or self._GetDefaultSasSSLCertPath(),
                            ssl_key or self._GetDefaultSasSSLKeyPath()))
 
+  def GetUrl(self,url, ssl_cert=None, ssl_key=None):
```
Use the   master version of sas.py since changes are    removed

jni2000 commented on src/harness/test_harness_objects.py:
```diff
@@ -0,0 +1,299 @@
+  def getGrantObject(self, grant_id):
+    try:
+      return self.grant_objects[grant_id]
+    except KeyError:
```
In case   of key error it suppose to fail hence removing the exception handling so that   it will fail when exception occurs

jni2000 commented on src/harness/test_harness_objects.py:
```diff
@@ -0,0 +1,299 @@
+    # Concatenate all the heartbeat request of CBSD into single request
+    heartbeat_requests = []
+    for cbsd_object_item in self.cbsd_objects.values():
+      heartbeat_requests = cbsd_object_item.constructHeartbeatRequestForAllActiveGrants()
```
Thanks   for pointing out     All the constructed heart beat requests are appended

jni2000 commented on src/harness/test_harness_objects.py:
```diff
@@ -0,0 +1,299 @@
+    except KeyError:
+      logging.debug("CBSD Object not found in the Domain Proxy object,cbsd_id=%s.", cbsd_id)
+
+  def heartbeatRequestForAllGrants(self):
```
Changed   the name as per suggestion

jni2000 commented on src/harness/test_harness_objects.py:
```diff
@@ -0,0 +1,299 @@
+        except KeyError:
+          logging.debug("CBSD or the Grant object not found")
+
+        if heartbeat_response['response']['responseCode'] == 0:
```
Checking   the success heart beat response is moved to     mapResponseCodeToGrantState

jni2000 commented on src/harness/test_harness_objects.py:
```diff
@@ -0,0 +1,299 @@
+        logging.debug("CBSD or Grant object not found")
+
+      # Perform Heartbeat request for all grants
+      _,_ =self.heartbeatRequestForAllGrants()
```
Updated   as per suggestion

jni2000 commented on src/harness/test_harness_objects.py:
```diff
@@ -0,0 +1,299 @@
+    """
+    cbsd_objects = []
+    for cbsd_object in self.cbsd_objects.values():
+      if cbsd_object.hasActiveGrant():
```
Yes it   has to check for authorized hence changed

jni2000 commented on src/harness/test_harness_objects.py:
```diff
@@ -0,0 +1,299 @@
+    self.cbsd_objects = {}
+    self.testcase = testcase
+
+  def initialize(self, registration_requests, grant_requests):
```
Updated   as per suggestion

jni2000 commented on src/harness/test_harness_objects.py:
```diff
@@ -0,0 +1,299 @@
+                                           grant_object)
+      return heartbeat_requests, heartbeat_responses
+
+  def mapResponseCodeToGrantState(self, heartbeat_response_code, grant_object):
```
Updated   as per suggestion

jni2000 commented on src/harness/test_harness_objects.py:
```diff
@@ -0,0 +1,299 @@
+    """Maps the heartbeat response and update the grant state"""
+    if heartbeat_response_code == 502:
+      grant_object.is_terminated = True
+    elif heartbeat_response_code == 500:
```
Used the   enum class  for Grant response codes

jni2000 commented on src/harness/test_harness_objects.py:
```diff
@@ -0,0 +1,299 @@
+      grant_object.is_terminated = True
+      grant_object.authorized_in_last_heartbeat = False
+    elif heartbeat_response_code == 501:
+      grant_object.authorized_in_last_heartbeat = False
```
added   else case to log unknown error codes

jni2000 commented on src/harness/test_harness_objects.py:
```diff
@@ -0,0 +1,299 @@
+            }
+            grant_requests.append(grant_request)
+
+          del self.cbsd_objects[heartbeat_response['cbsdId']]
```
Agreed   we need to delete grant object

jni2000 commented:
Taran

We have addressed all your latest review comments, please take a look and let us know if it is good to move froward for approval. Thanks.

James

TaranFurnell commented on src/harness/test_harness_objects.py:
```diff
@@ -0,0 +1,319 @@
+                                        self.ssl_cert, self.ssl_key)['relinquishmentResponse']
+      for relinquish_response in relinquish_responses:
+        if relinquish_response['response']['responseCode'] != 0:
+          logging.error('Relinquishment request for the grantId= %s failed',relinquish_response['grantId'])
```
Not just log an error, raise an exception. The primary reason a relinquishment could fail is if the test is sending invalid relinquishment requests (in which case the test result is invalid and the bug should be fixed), otherwise something has gone horribly wrong and the SAS UUT is inconsistent with the state expected by the test (in which case any subsequent requests have unpredictable responses, and it could be an error in either the SAS UUT or the test code).

jni2000 replied:
Agreed. We have added assert statements to verify that the   relinquishment is successful.

jni2000 commented:
Kate and Taran

We have address your last review comments and added necessary assertions. Please take a look and let us know if we can move on.

kate-harrison commented on src/harness/test_harness_objects.py:
```diff
@@ -0,0 +1,324 @@
+  def getCbsdObjectById(self, cbsd_id):
+      return self.cbsd_objects[cbsd_id]
+
+  def heartbeatRequestForAllActiveGrants(self):
```
`heartbeatForAllActiveGrants` would be more accurate.

jni2000 replied:
Changed   as per suggestion.

kate-harrison commented on src/harness/test_harness_objects.py:
```diff
@@ -0,0 +1,324 @@
+    """Performs heartbeat request for all grants of CBSD devices at once.
+
+    Returns:
+      heartbeat_requests : list of heart beat requests which are constructed for all the grants
```
`heartbeat` is one word.

jni2000 replied:
Corrected   as per suggestion.

kate-harrison commented on src/harness/test_harness_objects.py:
```diff
@@ -0,0 +1,324 @@
+      heartbeat_requests_wrapped = {
+        'heartbeatRequest': heartbeat_requests
+      }
+      print heartbeat_requests_wrapped
```
Please remove debug statements.

jni2000 replied:
Removed   as per suggestion.

kate-harrison commented on src/harness/test_harness_objects.py:
```diff
@@ -0,0 +1,324 @@
+        heartbeat_requests_wrapped, self.ssl_cert, self.ssl_key)['heartbeatResponse']
+
+      for heartbeat_response in heartbeat_responses:
+        cbsd_object = self.cbsd_objects[heartbeat_response['cbsdId']]
```
It's safer to use the request instead of the response (in case there was an error with the HBT).

jni2000 replied:
Changed   as per suggestion.

TaranFurnell replied:
This change doesn't appear to have been done.

kate-harrison commented on src/harness/test_harness_objects.py:
```diff
@@ -0,0 +1,324 @@
+    """Constrcut unwrapped version of grant request."""
+    grant_request = {
+      'cbsdId': heartbeat_request['cbsdId'],
+      'operationParam': {
```
These aren't params of the heartbeat request. They are only in the heartbeat response.

jni2000 replied:
Thanks   for pointing out.     Changed as per suggestion.

TaranFurnell commented on src/harness/test_harness_objects.py:
```diff
@@ -0,0 +1,324 @@
+    # Checking if the number of registration requests matches number of grant requests
+    # There should be exactly one grant request per registration request.
+    assert len(grant_requests) == len(registration_requests)
+    cbsd_ids = self.testcase.assertRegistered(registration_requests)
```
This doesn't use the domain proxy certificate, please modify the assertRegistered function to allow an optional certificate and key then pass that here.

jni2000 replied:
Changed   as per suggestion.

TaranFurnell commented on src/harness/test_harness_objects.py:
```diff
@@ -0,0 +1,324 @@
+
+    # Make one CBSD object per successful grant request
+    for index, grant_request in enumerate(grant_requests):
+      grant_request["cbsdId"] = cbsd_ids[index]
```
cbsdId should be set before requesting any grants

jni2000 replied:
Changed   as per suggestion.

kate-harrison commented on src/harness/test_harness_objects.py:
```diff
@@ -0,0 +1,324 @@
+    else:
+      logging.error('Unknown code=%s received in Heartbeat Response',heartbeat_response_code)
+
+  def _constructGrantRequests(self,heartbeat_request):
```
The name should be singular since you are constructing only one.

jni2000 replied:
Corrected   as per suggestion.

TaranFurnell commented on src/harness/test_harness_objects.py:
```diff
@@ -0,0 +1,324 @@
+    # Make one CBSD object per successful grant request
+    for index, grant_request in enumerate(grant_requests):
+      grant_request["cbsdId"] = cbsd_ids[index]
+      if grant_responses[index]['response']['responseCode'] == 0:
```
Please compare to enum instead of magic number.

jni2000 replied:
Changed   as per suggestion.

kate-harrison commented on src/harness/test_harness_objects.py:
```diff
@@ -0,0 +1,324 @@
+                                                 self.ssl_cert, self.ssl_key)['grantResponse']
+
+      for grant_response, grant_request in zip(grant_responses, grant_requests):
+        cbsd_object = self.cbsd_objects[grant_request['cbsdId']]
```
What happens if the Grant request fails?

jni2000 replied:
We ignore the grants failure per Test dev F2F working doc. Do we need adding handling logic ?

kate-harrison replied:
Yes, because the next line will raise an exception (invalid key) if the grant request failed. So code *should* be ignoring the failure, but the current code is not actually ignoring it.

jni2000 replied:
Changed   as per suggestion.

TaranFurnell commented on src/harness/test_harness_objects.py:
```diff
@@ -0,0 +1,324 @@
+           else:
+             logging.error('Invalid Response since operation param not present in case of Terminated Grant')
+
+        elif heartbeat_response['response']['responseCode'] != 0:
```
Compare to enum, please check against SUSPENDED_GRANT not the absence of SUCCESS.

jni2000 replied:
Changed   as per suggestion.

TaranFurnell commented on src/harness/test_harness_objects.py:
```diff
@@ -0,0 +1,324 @@
+      grant_object.is_terminated = True
+    elif heartbeat_response_code == Grant.ResponseCodes.SUSPENDED_GRANT.value:
+      grant_object.authorized_in_last_heartbeat = False
+    elif heartbeat_response_code == Grant.ResponseCodes.UNSYNC_OP_PARAM.value:
```
Remove UNSYNC_OP_PARAM check, just treat it as an unknown code error. If an UNSYNC_OP_PARAM actually occurs we don't know the state of the SAS.

jni2000 replied:
Removed   as per suggestion.

TaranFurnell commented on src/harness/test_harness_objects.py:
```diff
@@ -0,0 +1,324 @@
+      grant_object.authorized_in_last_heartbeat = False
+      logging.error('Received UNSYNC_OP_PARAM code in Heartbeat Response')
+    else:
+      logging.error('Unknown code=%s received in Heartbeat Response',heartbeat_response_code)
```
Please also raise an exception, this should never happen during testing.

jni2000 replied:
Raised   exception incase of heartbeat response unknown error code.

TaranFurnell commented on src/harness/test_harness_objects.py:
```diff
@@ -0,0 +1,324 @@
+      logging.error('Unknown code=%s received in Heartbeat Response',heartbeat_response_code)
+
+  def _constructGrantRequests(self,heartbeat_request):
+    """Constrcut unwrapped version of grant request."""
```
nit: spelling, Construct

jni2000 replied:
Corrected   the typo.

TaranFurnell commented on src/harness/test_harness_objects.py:
```diff
@@ -0,0 +1,324 @@
+      # Validate the response of the relinquishment request
+      for relinquishment_request, relinquishment_response in \
+              zip(relinquishment_requests, relinquishment_responses):
+        self.testcase.assertEqual(relinquishment_response['response']['responseCode'], 0)
```
Assert with the enum not a magic number

jni2000 replied:
Used the   Relinquishment response code for SUCCESS

TaranFurnell commented on src/harness/test_harness_objects.py:
```diff
@@ -0,0 +1,324 @@
+    SUCCESS = 0
+    TERMINATED_GRANT = 500
+    SUSPENDED_GRANT = 501
+    UNSYNC_OP_PARAM = 502
```
Remove UNSYNC_OP_PARAM

jni2000 replied:
Removed   as per suggestion.

jni2000 commented on src/harness/test_harness_objects.py:
```diff
@@ -0,0 +1,324 @@
+  def getOperationParamsOfAllAuthorizedGrants(self):
+    """Returns the list of operation params of all authorized grants."""
+    operation_params = []
+    for grant_object in self.grant_objects.values():
```
Changed   as per suggestion.

jni2000 commented on src/harness/test_harness_objects.py:
```diff
@@ -0,0 +1,324 @@
+    for grant_object in self.grant_objects.values():
+      if grant_object.isGrantAuthorizedInLastHeartbeat():
+        operation_param = grant_object.getRequestOperationParam()
+        operation_params.append(operation_param)
```
Changed   as per suggestion.

jni2000 commented on src/harness/test_harness_objects.py:
```diff
@@ -0,0 +1,324 @@
+    for grant_object in self.grant_objects.itervalues():
+      if grant_object.isActive():
+        heartbeat_request = grant_object.constructHeartbeatRequest()
+        heartbeat_requests.append(heartbeat_request)
```
Changed   as per suggestion.

jni2000 commented:
Kate and Taran

We have responded to all your recent review comments and made necessary changes/fixes. There is only one comment for line 309 of tast_harness_objects.py from Kate. Our response is 

> "We ignore the grants failure per Test dev F2F working doc. Do we need adding handling logic ?"

We can add handling logic if required. let us know.

kate-harrison commented:
Confirmed, you can ignore grant request failures in this PR.

TaranFurnell commented on src/harness/test_harness_objects.py:
```diff
@@ -0,0 +1,327 @@
+                                              cert=self.ssl_cert, key=self.ssl_key)
+
+    # Copy the cbsdId from Registration response to grant requests.
+    for index, grant_request in enumerate(grant_requests):
```
Is there any reason not to use zip?
```python
for cbsd_id, grant_request in zip(cbsd_ids, grant_requests):
```

jni2000 replied:
We Reused the existing code as is.

Agreed we can use zip, thanks for suggestion.

TaranFurnell commented on src/harness/test_harness_objects.py:
```diff
@@ -0,0 +1,327 @@
+                                               self.ssl_cert, self.ssl_key)['grantResponse']
+
+    # Make one CBSD object per successful grant request.
+    for index, grant_request in enumerate(grant_requests):
```
zip takes an arbitrary number of arguments. Using it will make the code in this loop more readable.
```python
zip(grant_requests, grant_responses, registration_requests, cbsd_ids)
```

jni2000 replied:
Agreed we can use zip, thanks for suggestion.

TaranFurnell commented on src/harness/test_harness_objects.py:
```diff
@@ -0,0 +1,327 @@
+    """
+    # Concatenate all the heartbeat request of CBSD into single request.
+    heartbeat_requests = []
+    for cbsd_object_item in self.cbsd_objects.values():
```
itervalues.

jni2000 replied:
itervalues used instead of values.

TaranFurnell commented on src/harness/test_harness_objects.py:
```diff
@@ -0,0 +1,327 @@
+    returns: list of CBSD objects.
+    """
+    cbsd_objects = []
+    for cbsd_object in self.cbsd_objects.values():
```
itervalues.

jni2000 replied:
itervalues used instead of values.

TaranFurnell commented on src/harness/test_harness_objects.py:
```diff
@@ -0,0 +1,327 @@
+  # Define an enumeration class named ResponseCodes with members
+  # 'SUCCESS'
+  class RelinquishmentResponseCodes(Enum):
+    SUCCESS = 0
```
ResponseCodes already exists, no need to create a new enum just for relinquishment. Move ResponseCodes out of the Grant class and use that instead, arguably it should be in a separate file.

jni2000 replied:
Agreed we added response code in different file and used the same.

TaranFurnell commented on src/harness/test_harness_objects.py:
```diff
@@ -0,0 +1,327 @@
+             # Construct grant request for Terminated grants
+             grant_requests.append(self._constructGrantRequest(heartbeat_response))
+           else:
+             logging.error('Invalid Response since operation param not present in case of Terminated Grant')
```
Providing the grant ID and CBSD ID from the heartbeat request in this error message will help in understanding what happened if this case occurs.

jni2000 replied:
Changed as per suggestion

TaranFurnell commented on src/harness/test_harness_objects.py:
```diff
@@ -0,0 +1,327 @@
+            grant_requests.append(self._constructGrantRequest(heartbeat_request))
+
+            # Delete the grant object
+            del self.cbsd_objects[heartbeat_response['cbsdId']].grant_objects[heartbeat_response['grantId']]
```
From what I can tell this is only ever running for suspended grants, never terminated grants?

jni2000 replied:
Thanks for pointing out 
grant object will be deleted in case of terminated grant also.

jni2000 commented:
Taran

We have responded to your latest review comments and made all required changes/fixes, please take a look.

kate-harrison commented on src/harness/test_harness_objects.py:
```diff
@@ -0,0 +1,320 @@
+    relinquishment_requests = []
+    grant_requests = []
+
+    # Construct relinquishment and grant for the heart beat response which have operation param.
```
`heartbeat` is one word.

jni2000 replied:
fixed.

TaranFurnell commented on src/harness/test_harness_objects.py:
```diff
@@ -0,0 +1,320 @@
+      4. Heartbeat is performed again for all grants. Failures are ignored.
+    """
+    # Heartbeat on all active grants
+    heartbeat_requests = []
```
heartbeat_requests is assigned on the next line, remove this line.

jni2000 replied:
Removed   as per suggestion.

jni2000 commented:
Thanks, Kate and Taran. We will take a final check and address all the remaining new review comments.

jni2000 commented on src/harness/test_harness_objects.py:
```diff
@@ -0,0 +1,320 @@
+    self.cbsd_objects = {}
+    self.testcase = testcase
+
+  def addCbsdsAndGrants(self, registration_requests, grant_requests):
```
Changed   as per suggestion.

jni2000 commented on src/harness/test_harness_objects.py:
```diff
@@ -0,0 +1,320 @@
+
+    #perform grant operation.
+    grant_responses = self.testcase._sas.Grant(grant_requests_wrap,
+                                               self.ssl_cert, self.ssl_key)['grantResponse']
```
Added   check as per suggestion.

jni2000 commented on src/harness/test_harness_objects.py:
```diff
@@ -0,0 +1,320 @@
+      }
+      # Perform heartbeat requests.
+      heartbeat_responses = self.testcase._sas.Heartbeat(
+        heartbeat_requests_wrapped, self.ssl_cert, self.ssl_key)['heartbeatResponse']
```
Added   check as per suggestion.

jni2000 commented on src/harness/test_harness_objects.py:
```diff
@@ -0,0 +1,320 @@
+             logging.error('Invalid Response received for grantId=%s and cbsdId=%s',
+                           heartbeat_response['grantId'], heartbeat_response['cbsdId'])
+             # Delete the grant object
+             del self.cbsd_objects[heartbeat_response['cbsdId']].grant_objects[heartbeat_response['grantId']]
```
Changed   as per suggestion.

jni2000 commented:
Thanks to all reviewers. We have done all minor changes according to the latest review comments.
